MAINT: Refactor _new_sortlike and _new_argsortlike

jaimefrio opened the pull request:
Simplify the signature of both functions, remove code duplication
inside, add proper handling of byte-swapping for compound dtypes
(like complex, fixes #5441), and add functionality to deal with
dtypes containing Python object references. This last enhancement is
not needed for this PR, but supports subsequent ones.

jaimefrio commented:
There are at least two or three more large commits to complete this:

1. Create npy_aquicksort, npy_amergesort and npy_aheapsort functions in npy_sort, and modify PyArray_ArgSort and PyArray_ArgPartition to always call _new_argsortlike, passing these functions in needed.
2. Modify npy_quicksort, npy_mergesort and npy_heapsort functions in npy_sort, and modify PyArray_Sort and PyArray_Partition to always call _new_sortlike, passing these functions if needed.
3. Add a comprehensive set of tests checking all combinations of contiguous/non-contiguous, swapped/not-swapped, with ref/without refs are properly sorted, argsorted, partitioned and argpartitioned.

They could be made separately as different PRs, or added as new commits to this same PR, you get to choose.

njsmith commented:
I'm kinda busy running around like a chicken with its head cut off for
the next few weeks, so I probably won't review in any detail, but
multiple PRs is usually a good approach!

On Sat, Jan 17, 2015 at 12:17 AM, Jaime <notifications@github.com> wrote:
> There are at least two or three more large commits to complete this:
>
> Create npy_aquicksort, npy_amergesort and npy_aheapsort functions in
> npy_sort, and modify PyArray_ArgSort and PyArray_ArgPartition to always call
> _new_argsortlike, passing these functions in needed.
> Modify npy_quicksort, npy_mergesort and npy_heapsort functions in npy_sort,
> and modify PyArray_Sort and PyArray_Partition to always call _new_sortlike,
> passing these functions if needed.
> Add a comprehensive set of tests checking all combinations of
> contiguous/non-contiguous, swapped/not-swapped, with ref/without refs are
> properly sorted, argsorted, partitioned and argpartitioned.
>
> They could be made separately as different PRs, or added as new commits to
> this same PR, you get to choose.
>
> —
> Reply to this email directly or view it on GitHub.



-- 
Nathaniel J. Smith
Postdoctoral researcher - Informatics - University of Edinburgh
http://vorpus.org

jaimefrio commented on numpy/core/src/multiarray/item_selection.c:
```diff
@@ -787,255 +787,287 @@ PyArray_Choose(PyArrayObject *ip, PyObject *op, PyArrayObject *out,
+            npy_intp i;
+            for (i = 0; i < nkth; ++i) {
+                ret = part(bufptr, N, kth[i], pivots, &npiv, op);
+                if (ret < 0 || (hasrefs && PyErr_Occurred())) {
```
This line probably deserves some explaining... The way `dtype->f->compare` is set up for object dtype's, under Python3, when comparing two uncomparable objects it returns a 0, i.e. considers them to be equal, so in principle the sorting continues all the way, although the result will be crap. So after sorting each 1D array, if our type has refs and hence the GIL has not been released, we check whether an error was raised, and return early if it is the case.

jaimefrio commented on numpy/core/src/multiarray/item_selection.c:
```diff
@@ -787,255 +787,287 @@ PyArray_Choose(PyArrayObject *ip, PyObject *op, PyArrayObject *out,
+            bufptr = buffer;
+        }
+        else if (swap) {
+            /* If the array is swapped but contiguous, unswap in-place */
```
Hmmm... I just realized this optimization could leave a contiguous, swapped array, unwapped if there was an error in the sorting routine, which is not good, will push a new commit to fix it.

jaimefrio commented:
OK, I fixed the failing stuff, this should be ready for review now.

juliantaylor commented on numpy/core/src/multiarray/item_selection.c:
```diff
@@ -787,255 +787,303 @@ PyArray_Choose(PyArrayObject *ip, PyObject *op, PyArrayObject *out,
+            ret = sort(bufptr, N, op);
+            #if defined(NPY_PY3K)
+            /* Object comparisons may raise an exception in Python 3 */
+            if (hasrefs && PyErr_Occurred()) {
```
NPY_PY3K could be put into the if ()
also preprocessor code is usually not indented

juliantaylor replied:
hm its probably not defined in python2, nevermind then

juliantaylor replied:
or one could do a #if !PY3K #define PY3K 0, add three lines to remove 4 ;)
the indentation should be fixed regardless I think

juliantaylor commented on numpy/core/src/multiarray/item_selection.c:
```diff
@@ -787,255 +787,303 @@ PyArray_Choose(PyArrayObject *ip, PyObject *op, PyArrayObject *out,
+                                              it->dataptr, astride, N, elsize);
+                /* ...then swap in-place if needed */
+                if (swap) {
+                    copyswapn(valbuffer, elsize, NULL, 0, N, swap, op);
```
this block is the same as in sortlike, could be moved into a small function

jaimefrio replied:
I thought of it, but it would be a function with 9 or 10 parameters, unless we want to extract `hasrefs`, `copyswapn`, `elsize` and `swap` from `op` in each iteration.

juliantaylor replied:
hm right probably too many arguments for little code saved

juliantaylor commented:
looks good with some minor style suggestions

jaimefrio committed:
MAINT: Refactor _new_sortlike and _new_argsortlike

Simplify the signature of both functions, remove code duplication
inside, add proper handling of byte-swapping for compound dtypes
(like complex, fixes #5441), and add functionality to deal with
dtypes containing Python object references. This last enhancement is
not needed for this PR, but supports subsequent ones.

jaimefrio commented:
Ok, changed the indentatntion of the `#if`s, which I believe is what we settled on having to be changed.

juliantaylor commented:
alright, merging thanks
